cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -1336,41 +1376,309 @@ collectTxBodySimpleScripts TxBodyContent {
+      maybeToStrictMaybe $ Ledger.hashAuxiliaryData @StandardMary <$> auxData
+    txIns = [(fromShelleyTxIn input, ViewTx) | input <- toList inputs]
+    txOuts = fromTxOut ShelleyBasedEraMary <$> toList outputs
+    txFee = TxFeeExplicit TxFeesExplicitInMaryEra $ fromShelleyLovelace txfee
+    txValidityRange =
+      ( case invalidBefore of
+          SNothing -> TxValidityNoLowerBound
+          SJust s -> TxValidityLowerBound ValidityLowerBoundInMaryEra s
+      , case invalidHereafter of
+          SNothing -> TxValidityNoUpperBound ValidityNoUpperBoundInMaryEra
+          SJust s -> TxValidityUpperBound ValidityUpperBoundInMaryEra s
+      )
+    (txMetadata, txAuxScripts) =
+      case auxData of
+        Nothing -> (TxMetadataNone, TxAuxScriptsNone)
+        Just s ->
+          let (ms, ss) = fromMaryAuxiliaryData s
+          in  ( if null ms then
+                  TxMetadataNone
+                else
+                  TxMetadataInEra TxMetadataInMaryEra (TxMetadata ms)
+              , case ss of
+                  [] -> TxAuxScriptsNone
+                  _  -> TxAuxScripts AuxScriptsInMaryEra ss
+              )
+    txWithdrawals
+      | null (Shelley.unWdrl withdrawals) = TxWithdrawalsNone
+      | otherwise =
+          TxWithdrawals WithdrawalsInMaryEra $ fromShelleyWithdrawal withdrawals
+    txCertificates
+      | null certificates = TxCertificatesNone
+      | otherwise =
+          TxCertificates
+            CertificatesInMaryEra
+            (map fromShelleyCertificate $ toList certificates)
+            ViewTx
+    txUpdateProposal =
+      case update of
+        SNothing -> TxUpdateProposalNone
+        SJust p ->
+          TxUpdateProposal UpdateProposalInMaryEra $ fromShelleyUpdate p
+    txMintValue
+      | isZero mint = TxMintNone
+      | otherwise = TxMintValue MultiAssetInMaryEra (fromMaryValue mint) ViewTx
+
+
+toShelleyWithdrawal :: [(StakeAddress, Lovelace, build)]
+                    -> Shelley.Wdrl StandardCrypto
 toShelleyWithdrawal withdrawals =
     Shelley.Wdrl $
       Map.fromList
         [ (toShelleyStakeAddr stakeAddr, toShelleyLovelace value)
         | (stakeAddr, value, _) <- withdrawals ]
 
+
+fromShelleyWithdrawal ::  Shelley.Wdrl StandardCrypto
+                      ->  [ ( StakeAddress
+                            , Lovelace
+                            , BuildTxWith ViewTx (Witness WitCtxStake era)
+                            )
+                          ]
+fromShelleyWithdrawal (Shelley.Wdrl withdrawals) =
+  [ (fromShelleyStakeAddr stakeAddr, fromShelleyLovelace value, ViewTx)
+  | (stakeAddr, value) <- Map.assocs withdrawals
+  ]
+
+
 -- | In the Shelley era the auxiliary data consists only of the tx metadata
 toShelleyAuxiliaryData :: Map Word64 TxMetadataValue
                        -> Ledger.AuxiliaryData StandardShelley
 toShelleyAuxiliaryData m =
     Shelley.Metadata
       (toShelleyMetadata m)
 
+
+fromShelleyAuxiliaryData  :: Ledger.AuxiliaryData StandardShelley
```

Comment:
This can be parameterized with `ShelleyBasedEra era` to avoid repetition

Reply:
Which repetition? Could you name specific functions?

Reply:
Generalization of `fromAllegraAuxiliaryData` and `fromMaryAuxiliaryData` incurred such strange constraints:

```hs
fromAllegraBasedAuxiliaryData
  ::  forall era ledgerera
  .   ( ShelleyLedgerEra era ~ ledgerera
      , Ledger.AuxiliaryData ledgerera ~ ShelleyMA.AuxiliaryData ledgerera
      , FromCBOR (CBOR.Annotator (Ledger.Script ledgerera))
      , ToCBOR (Ledger.Script ledgerera)
      , Ord (Ledger.Script ledgerera)
      )
  =>  ShelleyBasedEra era
  ->  Ledger.AuxiliaryData (ShelleyLedgerEra era)
  ->  (Map Word64 TxMetadataValue, [ScriptInEra era])
fromAllegraBasedAuxiliaryData era (ShelleyMA.AuxiliaryData ms ss) =
  (fromShelleyMetadata ms, fromShelleyBasedScript era <$> toList ss)
```

Reply:
This is fine for now.

Reply:
Applied.